fix: use ERC20Permit and update constructor

SvenMeyer opened the pull request:
Since @openzeppelin/contracts v4.9 is only redirecting to the now final ERC20Permit implementatation.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v4.9/contracts/token/ERC20/extensions/draft-ERC20Permit.sol

Since @openzeppelin contracts v5.0 `draft-ERC20Permit.sol` has been removed.

SvenMeyer committed:
fix: use ERC20Permit and update constructor

SvenMeyer committed:
feat: do not prepend 'x' to token name and symbol

0xShaito commented:
Linked issues:

https://github.com/defi-wonderland/xERC20/issues/3

https://github.com/defi-wonderland/xERC20/issues/4

0xShaito reviewed (changes requested):
Hi @SvenMeyer. PR looks great and it's a good idea! For it to be ready we need:

1) The version of openzeppelin should be updated in the `.gitmodules`:

(we can safely remove the other part [submodule "lib\\openzeppelin-contracts"])

```
[submodule "lib/openzeppelin-contracts"]
	path = lib/openzeppelin-contracts
	url = https://github.com/OpenZeppelin/openzeppelin-contracts
	branch = merge/release-v5.0
```

2) This will require a change in the version of the compiler in `foundry.toml` and in the contracts (replace all):

```
[profile.default]
solc = '0.8.23'
```

3) Some small tests will fail like the one that checks the symbol!

After that the pr should be fully ready and we will merge it as soon as we try it :) 

ty!

SvenMeyer committed:
Merge branch 'defi-wonderland:main' into main

SvenMeyer committed:
test: do not prepend 'x' to token name and symbol

SvenMeyer committed:
build: remove openzeppelin-contracts

SvenMeyer committed:
build: install openzeppelin-contracts@release-v4.9

SvenMeyer committed:
build: upgrade to solc 0.8.23

SvenMeyer committed:
build: install OpenZeppelin/openzeppelin-contracts@release-v4.9

SvenMeyer committed:
fix: revert usage of Ownable to match OZ contracts v4.9

SvenMeyer committed:
test: do not expect token name and symbol to start with 'x'

SvenMeyer commented:
@0xShaito Sorry for the delay .. wanted to get it working first
- reverted to `OpenZeppelin/openzeppelin-contracts@release-v4.9` to not introduce any braking changes at this stage
- fixed tests to not expect token name and symbol to start with 'x' (after deployment, token name and symbol will be as specified = unchanged)
- cloned into a fresh repo, all tests working

Please review

```
$ forge install --no-commit OpenZeppelin/openzeppelin-contracts@release-v4.9
Installing openzeppelin-contracts in /XXXXXXXXXXX/xERC20/lib/openzeppelin-contracts (url: Some("https://github.com/OpenZeppelin/openzeppelin-contracts"), tag: Some("release-v4.9"))
    Installed openzeppelin-contracts release-v4.9

$ git submodule
 9310e879db8ba3ea6d5c6489a579118fd264a3f5 lib/ds-test (remotes/origin/windows-compat-15-g9310e87)
 9b49a72cfdb36bcf195eb863f868f01a6d6d3186 lib/forge-std (v1.5.5-3-g9b49a72)
 73ce8b9adaf951e3ec702493bd4da010d74eaa74 lib/isolmate (v6-107-g73ce8b9)
+9329cfacd4c7d20bcb43d772d947ff9e39b65df9 lib/openzeppelin-contracts (v4.9.3-6-g9329cfac)
 576f549a7351814f112edcc42f3f8472d1712673 lib/permit2 (0x000000000022D473030F116dDEE9F6B43aC78BA3-13-g576f549)
 ce5d128391feb6f64dff05d8d4431c799646b2c5 lib/prb-test (v0.1.1)
```

excaliborr reviewed (approved):
lgtm!

SvenMeyer commented:
@0xShaito @excaliborr @prathmeshkhandelwal1 Anything left to do from my side ?

0xGorilla commented:
@SvenMeyer I did replicate your changes on PR #7 since I needed to also deploy everything to 10 chains and target the dev branch.

Ty for your contribution!
